### PR TITLE
Infection model ready to be used (but not fully functional yet!)

### PR DESCRIPTION
**Pull request by eMaerthin**
You can take a look, but it is still early phase of the development :)

### CODE REVIEW

File: src/models/infection_model.py [hunk 1]
```diff
@@ -0,0 +1,747 @@
...
+    def affected_people(self):
+        return self._affected_people
+
+    @property
+    def deaths(self):
+        return self._deaths
+
+    def draw_expected_case_severity(self):
+        # age_min, age_max(exclusively), fatality_prob
+        age_induced_fatality_rates = [(0, 20, 0.002), (20, 40, 0.002), (40, 50, 0.004), (50, 60, 0.013),
```
Comment (onomatopeia) [inline src/models/infection_model.py]: Move to constants?

File: src/models/infection_model.py [hunk 2]
```diff
@@ -0,0 +1,747 @@
...
+                return np.random.poisson, [], {'lam': lambda_}
+            raise ValueError(f'Sampling from distribution {distribution} is not yet supported but we can quickly add it')
+        f, args, kwargs = cached_random_gen(**kwargs)
+        return f(*args, **kwargs)
+
+    def handle_t0(self, id):
+        if self._infection_status[id] == InfectionStatus.Contraction:
+            self._infection_status[id] = InfectionStatus.Infectious
+        elif self._infection_status[id] != InfectionStatus.Infectious:
+            logger.error(f'state machine failure. was {self._infection_status[id]} id: {id}')
```
Comment (onomatopeia) [inline src/models/infection_model.py]: Shouldn't you return?

File: src/models/infection_model.py [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+            ))
+            realizations = dis.rvs(size=len(self._df_individuals[cond]))
+            values = [keys[r] for r in realizations]
+            df = pd.DataFrame(values, index=self._df_individuals[cond].index)
+            d = {**d, **df.to_dict()[0]}
+        return d
+
+    @staticmethod
+    def generate_random_sample(**kwargs) -> float:
+        @lru_cache(maxsize=32)
```
Comment (onomatopeia) [inline src/models/infection_model.py]: Hmm... that looks interesting. Have you investigated time improvement with and without it? Just out of curiosity, I never used it.

File: src/models/infection_model.py [hunk 4]
```diff
@@ -0,0 +1,747 @@
...
+        def cached_random_gen(**kwargs):
+            distribution = kwargs.get('distribution', 'poisson')
```
Comment (onomatopeia) [inline src/models/infection_model.py]: I think you have a constant with poisson as a default distribution somewhere, it looks like a good place to use it ;)

File: src/models/infection_model.py [hunk 5]
```diff
@@ -0,0 +1,747 @@
...
+                if approximate_distribution == LOGNORMAL:
+                    shape, loc, scale = scipy.stats.lognorm.fit(array, floc=0)
+                    return scipy.stats.lognorm.rvs, [shape], {'loc': loc, 'scale': scale}
+                elif approximate_distribution == GAMMA:
+                    shape, loc, scale = scipy.stats.gamma.fit(array, floc=0)
+                    return scipy.stats.gamma.rvs, [shape], {'loc': loc, 'scale': scale}
+                if approximate_distribution:
+                    raise ValueError(f'Approximating to this distribution {approximate_distribution}'
+                                     f'is not yet supported but we can quickly add it if needed')
+                # TODO: support no approximate distribution provided
```
Comment (onomatopeia) [inline src/models/infection_model.py]: Perhaps you could raise ValueError (or NotImplementedError) for not approximate distributions, as they are not supported yet.

File: src/models/infection_model.py [hunk 6]
```diff
@@ -0,0 +1,747 @@
...
+
+            if distribution == LOGNORMAL:
+                mean = kwargs.get('mean', 0.0)
+                sigma = kwargs.get('sigma', 1.0)
+                return np.random.lognormal, [], {'mean': mean, 'sigma': sigma}
+
+            lambda_ = kwargs.get('lambda', 1.0)
+            if distribution == 'exponential':
```
Comment (onomatopeia) [inline src/models/infection_model.py]: Calls for a constant. Same with Poisson two lines below.

File: src/models/infection_model.py [hunk 7]
```diff
@@ -0,0 +1,747 @@
...
+                Schema(lambda x: os.path.exists(x)).validate(filepath)
+                array = np.load(filepath)
+                approximate_distribution = kwargs.get('approximate_distribution', None)
+                if approximate_distribution == LOGNORMAL:
+                    shape, loc, scale = scipy.stats.lognorm.fit(array, floc=0)
+                    return scipy.stats.lognorm.rvs, [shape], {'loc': loc, 'scale': scale}
+                elif approximate_distribution == GAMMA:
+                    shape, loc, scale = scipy.stats.gamma.fit(array, floc=0)
+                    return scipy.stats.gamma.rvs, [shape], {'loc': loc, 'scale': scale}
+                if approximate_distribution:
```
Comment (onomatopeia) [inline src/models/infection_model.py]: Readability improvement (i.e. to support reading code without using one's brain): Seems to be an inconsistent usage of if / elif. `if LOGNORMAL: return; if GAMMA: return; if approximate_...: raise` or `if LOGNORMAL: return; elif GAMMA: return; elif approximate_...: raise`.

File: src/models/infection_model.py [hunk 8]
```diff
@@ -0,0 +1,747 @@
...
+            return time_events_
+
+        import_intensity = self._params[IMPORT_INTENSITY]
+        f_choice = convert_import_intensity_functions(import_intensity[FUNCTION])
+        if f_choice == ImportIntensityFunctions.NoImport:
+            return
+        func = import_intensity_functions[f_choice]
+        multiplier = import_intensity[MULTIPLIER]
+        rate = import_intensity[RATE]
+        cap = import_intensity[CAP]
```
Comment (onomatopeia) [inline src/models/infection_model.py]: What's cap?

File: config.py [hunk 9]
```diff
@@ -0,0 +1,2 @@
+import os
+ROOT_DIR = os.path.abspath(os.curdir)
```
[the remark this answers is not in the document]
Comment (eMaerthin) [reply config.py]: sounds good, thanks for the advise!

File: src/models/enums.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+class InfectionStatus(EnumWithPublicValue2MemberMap):
+    Healthy = 'healthy'
+    Contraction = CONTRACTION
+    Infectious = INFECTIOUS
+    StayHome = 'stay_home'
+    #SeeDoctor = 'see_doctor'  # probable, dynamic variable - visiting doctor can cause spreading the disease into vulnerable people (already sick)
+    Hospital = 'hospital'  # probable
+    Recovered = 'recovered'  # probable
+    Death = 'death'  # probable, alternative to Recovered
```
[the remark this answers is not in the document]
Comment (eMaerthin) [reply src/models/enums.py]: completely agree. I am still not using `StateDependentOnTheEpidemicsState` - I plan to implement backtrace today, maybe I will implement it in a different branch. 
infectious = this is rather period from symptom onset until further action - so during this time the individual is mostly infectious, but recent literature suggest infectivity is even before time of the symptom onset unfortunately.

File: src/models/enums.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    This enum is intended for storing all initial condition selection algorithms,
+    to seed initially infectious agents in the population
+    RandomSelection assignes infection status randomly based on provided distribution of sick people
+    In the future other algorithms could be added e.g. to limit assigning infection status to people who work or commute
+    """
+    RandomSelection = 'random_selection'
+
+
+class KernelType(EnumWithPublicValue2MemberMap):
```
[the remark this answers is not in the document]
Comment (eMaerthin) [reply src/models/enums.py]: Sounds like a good enhancement! Let's remove `KernelType` class when your proposition is implemented, ok?

File: src/models/enums.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    Sporadic = SPORADIC
+    Household = HOUSEHOLD
+    Friendship = FRIENDSHIP
+    Workplace = WORKPLACE
+    Transport = TRANSPORT
+    Constant = CONSTANT
+
+
+class ImportIntensityFunctions(EnumWithPublicValue2MemberMap):
+    NoImport = NO_IMPORT
```
[the remark this answers is not in the document]
Comment (eMaerthin) [reply src/models/enums.py]: this list is for various ways of importing new infected cases into the population according to provided schedule.
`NoImport` means that we have a lockdown and no further cases will be imported from "nowhere"

File: src/models/enums.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    Exponential = EXPONENTIAL # a * exp(r*t)
+    Polynomial = POLYNOMIAL  # a * r ^ t
+
+
+class FearFunctions(EnumWithPublicValue2MemberMap):
+    """
```
[the remark this answers is not in the document]
Comment (eMaerthin) [reply src/models/enums.py]: yeah, I am happy to have this enhancement at some point. I will add this as a todo comment to code. Additionally, fear_function1 will be renamed

File: src/models/infection_model.py [hunk 14]
```diff
@@ -0,0 +1,747 @@
...
+        with open(params_path, 'r') as params_file:
+            params = json.loads(
+                params_file.read()
+            )  # TODO: check whether this should be moved to different place
+        self._params = dict()
+        self.event_queue = []
+        self._affected_people = 0
+        self._deaths = 0
+        logger.info('Parsing params...')
+        for key, schema in infection_model_schemas.items():
```
[the remark this answers is not in the document]
Comment (eMaerthin) [reply src/models/infection_model.py]: - this is defined in schemas.py 
- yes it is a constant

File: src/models/schemas.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    }),
+    FEAR_FACTORS: fear_factors_schema,
+    IMPORT_INTENSITY: import_intensity_schema,
+    START_TIME: global_time_schema,
+    DEATH_PROBABILITY: death_probability_schema,
+    RANDOM_SEED: random_seed_schema
+}
+
+
+def event_schema_fun(df_individuals):
```
[the remark this answers is not in the document]
Comment (eMaerthin) [reply src/models/schemas.py]: Hmm lets remove this schema. This is huge overhead for the simulation, therefore I was not using it anyway.

File: src/models/infection_model.py (hunk 2 above)
Comment (eMaerthin) [reply src/models/infection_model.py]: I will change the code to return AssertionError

File: src/models/infection_model.py (hunk 3 above)
Comment (eMaerthin) [reply src/models/infection_model.py]: haha actually after removing `@lru_cache` code became a little bit quicker, thanks for asking:)

File: src/models/infection_model.py (hunk 4 above)
Comment (eMaerthin) [reply src/models/infection_model.py]: good point

File: src/models/infection_model.py (hunk 8 above)
Comment (eMaerthin) [reply src/models/infection_model.py]: maximum number of event coming from that intensity function say - after cap=100 imported cases, borders will be closed and no further cases will be imported

File: src/models/infection_model.py (hunk 7 above)
Comment (eMaerthin) [reply src/models/infection_model.py]: thanks, I made it more consistent now.

File: src/models/enums.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (onomatopeia) [reply src/models/enums.py]: Agree. In case we are asked to extend the model it can be changed.

File: src/models/enums.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+
+class ImportIntensityFunctions(EnumWithPublicValue2MemberMap):
+    """
+    Here we list various ways of importing new infected cases into the population according to provided schedule.
+    - NoImport means that we have a lockdown already so no further cases will be imported from the outside.
+    - Exponential - there will be added N imported cases (for instance due to travels, migration)
+                    up to time T that solves the equation N=a*exp(rT)
+    - Polynomial - imported cases are drawn in polynomial rate
+                    - N imported cases up to time T that solves the equation N=a*pow(r, T)
+    """
```
Comment (onomatopeia) [inline src/models/enums.py]: Nice :)

File: src/models/enums.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+class SupportedDistributions(EnumWithPublicValue2MemberMap):
+    Lognormal = LOGNORMAL
+    Exponential = EXPONENTIAL
+    Poisson = POISSON
+    Gamma = GAMMA
+    FromFile = FROM_FILE
+
+
+def _convert_enum(enum_class, x):
+    for status in enum_class:
```
Comment (onomatopeia) [inline src/models/enums.py]: Do I understand correctly that given an enum, let's say `class Gender(Enum): MALE=0; FEMALE=1` and a value, say 1, you want to get the enum value, here Gender.FEMALE?
If so then:
```
>>> from enum import Enum
>>> class Gender(Enum):
...     MALE=0
...     FEMALE=1
...
>>> g = 1
>>> Gender(g)
<Gender.FEMALE: 1>
```

Comment (onomatopeia) [reply src/models/enums.py]:
```
>>> Gender(2)
ValueError: 2 is not a valid Gender

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\onomatopeia\.conda\envs\modelling-ncov2019\lib\enum.py", line 310, in __call__
    return cls.__new__(cls, value)
  File "C:\Users\onomatopeia\.conda\envs\modelling-ncov2019\lib\enum.py", line 564, in __new__
    raise exc
  File "C:\Users\onomatopeia\.conda\envs\modelling-ncov2019\lib\enum.py", line 548, in __new__
    result = cls._missing_(value)
  File "C:\Users\onomatopeia\.conda\envs\modelling-ncov2019\lib\enum.py", line 577, in _missing_
    raise ValueError("%r is not a valid %s" % (value, cls.__name__))
ValueError: 2 is not a valid Gender
```

File: src/models/enums.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (onomatopeia) [reply src/models/enums.py]: let's drop it for now

File: src/models/enums.py (hunk 17 above)
Comment (eMaerthin) [reply src/models/enums.py]: nice

**Pull request merged**